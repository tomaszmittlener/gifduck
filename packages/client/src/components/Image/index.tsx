import React, { FunctionComponent, useState, useEffect, useRef } from 'react'
import styled from 'styled-components'

interface SearchResultsProps {
  className?: string
  testId?: string
  src: string
  height: number
  width: number
  aspectRatio: number
  preSrc: string
}

const PLACEHOLDER =
  'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII='

const Container = styled.figure<{ aspectRatio: number }>`
  margin: 0;
  position: relative;
  width: 100%;
  height: 100%;
  &:before {
    display: block;
    content: '';
    width: 100%;
    padding-top: ${({ aspectRatio }) => aspectRatio}%;
  }
`

const StyledImage = styled.img<{ isImageLoaded: boolean }>`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  opacity: ${({ isImageLoaded }) => (isImageLoaded ? 1 : 0)};
`

const Image: FunctionComponent<SearchResultsProps> = ({ className, testId, src, aspectRatio, preSrc }) => {
  const containerRef = useRef<HTMLElement>(null)
  const [isImageLoaded, setImageLoaded] = useState<boolean>(false)

  const [preUrl, setPreUrl] = useState<string>(PLACEHOLDER)
  const [imgUrl, setImgUrl] = useState<string>('')

  const onImgLoad = () => {
    setImageLoaded(true)
  }

  const onImgError = () => {
    setImgUrl(PLACEHOLDER)
  }
  const onPreError = () => {
    setPreUrl(PLACEHOLDER)
  }

  // lazy load images
  useEffect(() => {
    let observer: IntersectionObserver
    let didCancel = false

    if (containerRef?.current && imgUrl !== src && preSrc !== preUrl) {
      observer = new IntersectionObserver(
        entries => {
          entries.forEach(entry => {
            if (containerRef?.current && !didCancel && (entry.intersectionRatio > 0 || entry.isIntersecting)) {
              setPreUrl(preSrc)
              setImgUrl(src)
              observer.unobserve(containerRef.current)
            }
          })
        },
        {
          threshold: 0.01,
          rootMargin: '75%',
        },
      )
      observer.observe(containerRef.current)
    }
    return () => {
      didCancel = true
      // on component cleanup, we remove the listener
      if (observer && observer.unobserve) {
        containerRef.current && observer.unobserve(containerRef.current)
      }
    }
  }, [src, imgUrl, preSrc, preUrl, containerRef])

  return (
    <Container data-testid={testId} className={className} aspectRatio={aspectRatio} ref={containerRef}>
      {/*Original Image - will appear when loaded. initially has no src - it will be added on lazyload*/}
      {imgUrl && (
        <StyledImage
          src={imgUrl}
          onLoad={onImgLoad}
          onError={onImgError}
          isImageLoaded={isImageLoaded}
          style={{ zIndex: 0 }}
        />
      )}
      {/*Preview Image - will appear on lazyload and disappear when original image is loaded*/}
      <StyledImage src={preUrl} onError={onPreError} isImageLoaded={!isImageLoaded} style={{ zIndex: 100 }} />
    </Container>
  )
}

export default Image
