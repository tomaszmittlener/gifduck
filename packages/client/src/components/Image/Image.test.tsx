import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import Theme from 'styles/theme'

import Image from './'

const defaultProps = {
  className: 'class-name',
  testId: 'test-id',
  src: 'image-url',
  height: 10,
  width: 20,
  aspectRatio: 1.22222,
  preSrc: 'pre-scr',
}

afterEach(cleanup)

describe('Image', () => {
  test('renders correctly', async () => {
    const { container } = render(
      <ThemeProvider theme={Theme}>
        <Image {...defaultProps} />
      </ThemeProvider>,
    )
    expect(container).toMatchSnapshot()
  })

  test('passes className to container', async () => {
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <Image {...defaultProps} />
      </ThemeProvider>,
    )

    expect(getByTestId(defaultProps.testId)).toHaveClass(defaultProps.className)
  })
})
