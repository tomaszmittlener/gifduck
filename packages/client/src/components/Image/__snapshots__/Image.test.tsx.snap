// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly 1`] = `
.c0 {
  margin: 0;
  position: relative;
  width: 100%;
  height: 100%;
}

.c0:before {
  display: block;
  content: '';
  width: 100%;
  padding-top: 20%;
}

.c1 {
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  opacity: 1;
}

<div>
  <figure
    class="c0 class-name"
    data-testid="test-id"
  >
    
    <img
      class="c1"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
      style="z-index: 100;"
    />
  </figure>
</div>
`;
